File: app/src/main/java/com/rkopylknu/minimaltodo/Utility/TodoNotificationService.java
package com.rkopylknu.minimaltodo.Utility;

import android.app.IntentService;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.os.Build;

import androidx.core.app.NotificationCompat;
import androidx.core.content.ContextCompat;

import com.rkopylknu.minimaltodo.R;
import com.rkopylknu.minimaltodo.Reminder.ReminderActivity;

import java.util.UUID;

public class TodoNotificationService extends IntentService {
    public static final String TODOTEXT = "com.avjindersekhon.todonotificationservicetext";
    public static final String TODOUUID = "com.avjindersekhon.todonotificationserviceuuid";
    private static final String CHANNEL_ID = "com.rkopylknu.minimaltodo";
    private String mTodoText;
    private UUID mTodoUUID;
    private Context mContext;

    public TodoNotificationService() {
        super("TodoNotificationService");
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        mTodoText = intent.getStringExtra(TODOTEXT);
        mTodoUUID = (UUID) intent.getSerializableExtra(TODOUUID);

        Intent contentIntent = new Intent(this, ReminderActivity.class);
        contentIntent.putExtra(TODOUUID, mTodoUUID);
        PendingIntent pendingContentIntent =
                PendingIntent.getActivity(
                        this,
                        mTodoUUID.hashCode(),
                        contentIntent,
                        PendingIntent.FLAG_UPDATE_CURRENT
                );

        Intent deleteIntent = new Intent(this, DeleteNotificationService.class);
        deleteIntent.putExtra(TODOUUID, mTodoUUID);
        PendingIntent pendingDeleteIntent =
                PendingIntent.getService(
                        this,
                        mTodoUUID.hashCode(),
                        deleteIntent,
                        PendingIntent.FLAG_UPDATE_CURRENT
                );

        createNotificationChannel();
        Notification notification = new NotificationCompat.Builder(this, CHANNEL_ID)
                .setContentTitle(mTodoText)
                .setContentText("")
                .setSmallIcon(R.drawable.ic_done_white_24dp)
                .setContentIntent(pendingContentIntent)
                .setDeleteIntent(pendingDeleteIntent)
                .setAutoCancel(true)
                .build();

        NotificationManager manager = ContextCompat.getSystemService(this, NotificationManager.class);
        manager.notify(mTodoUUID.hashCode(), notification);

        /*
        Log.d("OskarSchindler", "onHandleIntent called");
        NotificationManager manager = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
        Intent i = new Intent(this, ReminderActivity.class);
        i.putExtra(TodoNotificationService.TODOUUID, mTodoUUID);
        Intent deleteIntent = new Intent(this, DeleteNotificationService.class);
        deleteIntent.putExtra(TODOUUID, mTodoUUID);
        Notification notification = new Notification.Builder(this)
                .setContentTitle(mTodoText)
                .setSmallIcon(R.drawable.ic_done_white_24dp)
                .setDefaults(Notification.DEFAULT_SOUND)
                .setDeleteIntent(PendingIntent.getService(this, mTodoUUID.hashCode(), deleteIntent, PendingIntent.FLAG_UPDATE_CURRENT))
                .setContentIntent(PendingIntent.getActivity(this, mTodoUUID.hashCode(), i, PendingIntent.FLAG_UPDATE_CURRENT))
                .build();

        manager.notify(100, notification);
        */

//        Uri defaultRingone = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION);
//        MediaPlayer mp = new MediaPlayer();
//        try{
//            mp.setDataSource(this, defaultRingone);
//            mp.setAudioStreamType(AudioManager.STREAM_NOTIFICATION);
//            mp.prepare();
//            mp.setOnCompletionListener(new MediaPlayer.OnCompletionListener() {
//                @Override
//                public void onCompletion(MediaPlayer mp) {
//                    mp.release();
//                }
//            });
//            mp.start();
//
//        }
//        catch (Exception e){
//            e.printStackTrace();
//        }

    }

    private void createNotificationChannel() {
        // Create the NotificationChannel, but only on API 26+ because
        // the NotificationChannel class is new and not in the support library
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            int importance = NotificationManager.IMPORTANCE_DEFAULT;
            NotificationChannel channel =
                    new NotificationChannel(CHANNEL_ID, "Channel Name", importance);
            // Register the channel with the system; you can't change the importance
            // or other notification behaviors after this
            NotificationManager notificationManager = ContextCompat.getSystemService(this, NotificationManager.class);
            notificationManager.createNotificationChannel(channel);
        }
    }
}

File: app/src/main/java/com/rkopylknu/minimaltodo/Utility/TodoNotificationService.kt
package com.rkopylknu.minimaltodo.Utility

import android.app.IntentService
import android.content.Intent
import com.rkopylknu.minimaltodo.R
import com.rkopylknu.minimaltodo.Reminder.ReminderActivity
import android.app.PendingIntent
import androidx.core.app.NotificationCompat
import android.app.NotificationManager
import androidx.core.content.ContextCompat
import android.os.Build
import android.app.NotificationChannel
import android.content.Context
import java.util.*

class TodoNotificationService : IntentService("TodoNotificationService") {

    private var mTodoText: String? = null
    private var mTodoUUID: UUID? = null

    override fun onHandleIntent(intent: Intent?) {
        mTodoText = intent!!.getStringExtra(TODOTEXT)
        mTodoUUID = intent.getSerializableExtra(TODOUUID) as UUID?
        val contentIntent = Intent(this, ReminderActivity::class.java)
        contentIntent.putExtra(TODOUUID, mTodoUUID)

        val pendingContentIntent = PendingIntent.getActivity(
            this,
            mTodoUUID.hashCode(),
            contentIntent,
            PendingIntent.FLAG_UPDATE_CURRENT
        )

        val deleteIntent = Intent(this, DeleteNotificationService::class.java)
        deleteIntent.putExtra(TODOUUID, mTodoUUID)
        val pendingDeleteIntent = PendingIntent.getService(
            this,
            mTodoUUID.hashCode(),
            deleteIntent,
            PendingIntent.FLAG_UPDATE_CURRENT
        )

        createNotificationChannel()
        val notification = NotificationCompat.Builder(this, CHANNEL_ID)
            .setContentTitle(mTodoText)
            .setContentText("")
            .setSmallIcon(R.drawable.ic_done_white_24dp)
            .setContentIntent(pendingContentIntent)
            .setDeleteIntent(pendingDeleteIntent)
            .setAutoCancel(true)
            .build()
        val manager = ContextCompat.getSystemService(this, NotificationManager::class.java)
        manager!!.notify(mTodoUUID.hashCode(), notification)
    }

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val importance = NotificationManager.IMPORTANCE_DEFAULT
            val channel = NotificationChannel(CHANNEL_ID, "Channel Name", importance)
            val notificationManager =
                ContextCompat.getSystemService(this, NotificationManager::class.java)
            notificationManager!!.createNotificationChannel(channel)
        }
    }

    companion object {
        const val TODOTEXT = "com.avjindersekhon.todonotificationservicetext"
        const val TODOUUID = "com.avjindersekhon.todonotificationserviceuuid"
        private const val CHANNEL_ID = "com.rkopylknu.minimaltodo"
    }
}